Improvements for wxSTC autocompletion and call tips

NPDevelopmentCode opened the pull request:
As I mentioned on the dev list, I realize this is a large pull request; but I think the net result is a better behaving popup class for autocompletion lists and call tips.  This should also address all outstanding issues related to autocompletion.

- Commits 1 and 2 are additions to the stc sample.  The first adds a demonstration of autocompletion for C preprocessor directives.  The second demonstrates handling clicks in call tips by using 3 related call tips and letting the user cycle between the tips by clicking up or down buttons.

![newcalltip](https://user-images.githubusercontent.com/10080894/54586626-a1ab2180-49eb-11e9-978d-1d01661e004f.png)

- Commit 3 adds the wxSTCPopupWindow class with customizations for windows and GTK+ to make sure it has the behavior it needs.  The class is based on the popup class currently used so on any other port the behavior should be the same as it is now.

- The next four commits (4 through 7) implement the wxSTCPopupWindow for wxCocoa.  The first of these adds an objective C file with several functions needed for the popup class and uses them in the C++ code.  Since I had to add a new source file and header, the next commit updates the wxWidgets build system files to use the new files.  Commit 6 is the result of running upmake and bakefile.  The final commit of this set adds the new file to the xcode project.  This is obviously unfotunate, but I could not find any other way to get a popup to not take focus on macos any other way.

- Commits 8 and 9 use the new popup window to show autocompletion lists and call tips respectively.

- Commit 10 introduces a custom listbox class based on wxVListBox.  The custom listbox is built specifically to never take focus but always look like it has focus.  This commit also adds a helper class to keep track of the listbox’s images and colors.

- Commits 11 through 15 are enhancements for the popup classes. 

  * Commits 11 and 12 let a popup window move when its parent stc window moves and hide when its application window is minimized.

  * Commit 13 is windows specific and is intended to reduce flicker when switching between different call tips.  I had to switch the drawing behavior to cache the calltip background in a bitmap and draw the bitmap in the erase background event and add a very quick fade in effect for call tip windows.  I would have thought either of those would have been sufficient, but I had to do both to get rid of the flicker.

  * Commit 14 adds a method to wxSTC to let the colors be specified for an autocompletion window. Commit 15 adds a method to make the autocompletion list look like it’s being shown with a list control instead of a listbox.  I added this because that’s how autocompletion lists are currently shown, and this lets a user maintain the current style if they want to.  Also, I think on windows it just looks better. 

![autocomplistctrl](https://user-images.githubusercontent.com/10080894/54586636-a96ac600-49eb-11e9-87e1-a8bbf881edac.png)

- Commit 16 is just regenerating the stc files after changes where made stc.cpp.in and stc.h.in.

- Commits 17 through 19 add a gui test to verify the behavior of the stc popups is correct.  Commit 17 adds the test file and modifies the build system files.  Commit 18 is the result of running bakefile.  The final commit updates the remaining build files to link with the stc library.

I hope the breakdown of commits is helpful and I won't be offended if it takes a while to evaluate this pull request.

Commit:
Demonstrate autocompletion in the stc sample

This adds a demonstration of autocompletion for C preprocessor
directives to the stc sample. It also shows how to register and use
small images with the autocompletion popup.

Commit:
Allow wxSTCPopupWindow to move when its parent wxSTC moves

Commit:
Demonstrate call tip clicks in the stc sample

This adds a demonstration of handling call tip clicks to the stc sample.
Instead of showing a single call tip, 3 different call tips can be
shown. The sample demonstrates how to move between the call tips
depending on if the up or down button was clicked in the call tip
window.

Commit:
Modify xcode project to use PlatWXcocoa.mm

Commit:
Use wxSTCPopupWindow for call tips in wxSTC

This class has been built to have the features needed for showing popups
with wxSTC and so provides a much shorter implementation for calltips.

Commit:
Reduce flicker with STC call tips in MSW

When switching between call tips in MSW, there can be a bit of flicker
when the first is closed and the new one is opened. To reduce the
flicker, store the call tip background in a bitmap and use a very brief
fade-in animation when showing the new call tip.

Commit:
Update other test build files by hand for the new STC test

Commit:
Add wxSTCPopupWindow for wxSTC

The wxSTCPopupWindow is intended to serve as a popup window for showing
call tips and popup lists with wxSTC. This class is designed to show its
content and respond to mouse clicks, but it should never take focus from
its parent wxSTC. It is built with customizations for wxMSW, wxCocoa,
and wxGTK+ to function in that way.

Commit:
Hide popup when STC app is minimized on macos and GTK+

Commit:
Define the wxSTCPopupBase for wxCocoa

With the cocoa port, wxSTCPopupBase is defined by creating a floating
window using the cocoa api and then wrapping that window in a
wxNonOwnedWindow for use with wxWidgets.

Commit:
Use wxSTCPopupWindow for showing autocompletion lists

The newly defined wxSTCPopupWindow class has been built to have the
behavior necessary for showing the popup windows used by wxSTC. This
commit uses it as the base class of the window that shows autocompletion
lists.

Commit:
Implement wxSTCListBox using wxVListBox

Previously wxSTCListBox was a class derived from wxListView and required
several extra steps to make the control look correct when it lacked
focus. This commit changes wxSTCListBox so that it is a wxVListBox and
is built to always draw itself looking as though it has focus.

In addition this splits the work previously done by ListBoxImpl class
among wxSTCListBox and a newly defined wxSTCListBoxVisualData class.
wxSTCListBox manages the work done by a specific list and
wxSTCListBoxVisualData manages the data common to all lists. All
ListBoxImpl methods now simply forward to a method of one of those 2
classes.

Commit:
Add AutoCompSetColours method to stc

With some system themes, the default colors used for a list box can be
hard to read or otherwise unsuitable for use with an autocompletion
popup. This method lets a user manually specify colours for the list
box’s background, text, selection background, and selected text.

Commit:
Add AutoCompUseListCtrl method to STC

The AutoCompUseListCtrl method can be used to make an autocompletion
list look like it's being shown with a wxListCtrl instead of a
wxListBox. When this style is used, the list will have hot tracking.
On MSW, the colours will also be slightly different.

Commit:
Regenerate STC files after recent changes

NewPagodi commented:
Oops.  I guess I need to add the wxScintilla lib for the gui test with the builds using static libs.  I need to ask for help since I don't know conditionally add that lib for static builds.

vadz commented:
Sorry, what's the actual problem? I see some compilation errors due to missing headers in Travis, but I'm not sure what's wrong with the libraries?

Commit:
Modify build system files to use PlatWXcocoa.mm

A recent commit added the file src/stc/PlatWXcocoa.mm needed for wxSTC
with the cocoa port. This commit modifies the build system files to
use this new file when it is needed.

Commit:
Regenerate build files after recent changes

NewPagodi commented:
4 of the build failures were due to some mistakes I made with the CMake files.  I temporarily removed the commits with the gui test so I could focus on those.  
Now that I the CMake issues fixed, I'll try to get the test working tomorrow.

Commit:
Add gui test for STC popup items

With wxSTC, popup autocompletion lists and call tips need to be able to
show their information and respond to mouse clicks, but should never
take focus from their parent STC. This test verifies that these popups
function in this manner.

Commit:
Regenerate the test make files to include the new STC test

NewPagodi commented:
I got the test working.  I had to make 2 changes.  First I surrounded the test with a check for `WXUSINGDLL` as discussed above.  Unfortunately, that means the test will only be run with the builds using shared libraries, but that's better than nothing.

Second, to get the test working with the 2 Travis cases where `--disable-stc` is set, I changed the definition of `WXLIB_STC` in common.bkl from

`<if cond="MONOLITHIC=='0'">$(mk.evalExpr(wxwin.mkLibName('stc')))</if>`

to

`<if cond="MONOLITHIC=='0' and USE_STC=='1'">$(mk.evalExpr(wxwin.mkLibName('stc')))</if>`

and then in test.bkl, I listed the link library as `<sys-lib>$(WXLIB_STC)</sys-lib>` as is currently done for the webview library.  That resulted in a change to autoconf_inc.m4 and the contributing docs say to run autoconf if that happens.  I did that, but I'm only following the directions and don't really understand what that does.

In short, I think this is right, but the second to last commit should probably be looked at closely.

vadz commented:
Thanks again for all your work on this!

I'm afraid I don't really understand why does the test only in the DLL builds, but I agree that it's not a critical problem. As for `WXLIB_STC` change, it definitely seems correct to me as we're already doing the same thing for the other libraries.

I'll try to review this, but it will be a cursory review at most and I don't expect to find any real problems, so this could be merged soon. However I'd like to keep it open for at least a while (a week?) and hope that people using wxSTC in their applications could test this PR and leave their comments if they find any problems.

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
+    wxSTCPopupBase::wxSTCPopupBase(wxWindow* parent):wxFrame()
+    {
+        // Make sure the frame is initially hidden. However, GTK+ will hide the
+        // frame initially and doesn't like trying to hide it before it's
```
This is pretty bad actually, hiding any window before creating it is definitely supposed to work.

OTOH frames are also supposed to be created hidden under all platforms, so this call shouldn't be _necessary_ anywhere. It should't do any harm however neither...

NewPagodi replied:
I just removed the call to Hide everywhere since I now know it's done by default everywhere.

vadz commented on samples/stc/edit.cpp:
```diff
@@ -169,6 +186,14 @@ Edit::Edit (wxWindow *parent, wxWindowID id,
+    wxBitmap bmp(hashtag_xpm);
+    RegisterImage(0, bmp);
+
+    //call tips
```
Very nitpicky:
```suggestion
    // call tips
```

vadz commented on samples/stc/edit.cpp:
```diff
@@ -481,11 +504,50 @@ void Edit::OnCharAdded (wxStyledTextEvent &event) {
+
+void Edit::OnCallTipClick(wxStyledTextEvent &event)
+{
+    if ( event.GetPosition() == 1 ) {
```
```suggestion
    if ( event.GetPosition() == 1 )
    {
```
(also in a few more places below)

NewPagodi replied:
This was to match the code style in the surrounding methods. Should I change it to match the current code guidelines?

NewPagodi replied:
I'll have to make a new commit to address the other 3 comments.  Should I change these at that time as well or should I leave them to match the surrounding methods?

vadz replied:
I haven't realized this was to match the existing style. I don't really know what to do in such cases, perhaps the best would be to commit this as is and then reformat it all at once in a later commit? In any case, please feel free to ignore this suggestion.

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
-};
+        // Do not activate in response to mouse clicks on this window.
+        bool wxSTCPopupBase::MSWHandleMessage(WXLRESULT *res, WXUINT msg,
+                                              WXWPARAM wParam, WXLPARAM lParam)
```
```suggestion
                                              WXWPARAM wParam, WXLPARAM lParam) wxOVERRIDE
```

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
-private:
-    wxDECLARE_EVENT_TABLE();
+        // Use ShowWithoutActivating instead of show.
+        bool wxSTCPopupBase::Show(bool show)
```
```suggestion
        bool wxSTCPopupBase::Show(bool show) wxOVERRIDE
```

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
+}
 
-    wxListView* GetLB() { return lv; }
+void SetColourHelper(bool& isSet, wxColour& itemCol, const wxColour& newColour)
```
```suggestion
static void SetColourHelper(bool& isSet, wxColour& itemCol, const wxColour& newColour)
```
(or put it an unnamed namespace)

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
+    wxColour m_textColour;
+    wxColour m_highlightBgColour;
+    wxColour m_highlightTextColour;
+    bool     m_bgColourIsSet;
```
Why do we need these `xxxIsSet` variables? It seems strange to store the colour validity outside of it when we could just use `xxx.IsOk()` instead.

NewPagodi replied:
The IsSet variables are tracking when the colors are set by the user.  An, admittedly exotic, example where they might be needed is in responding to system color changes.  When the `ComputeColours` function gets called, we need to recompute all the colors that have not been specifically set by the user.  At that point it won't be sufficient to check if the color is OK because all the colors will be OK due to a previous to the `ComputeColours` function.

I know there were other cases where those variables were needed, but that's the only one I remember right now.  I agree they're ugly and I tried to get rid of them, but I couldn't get everything to work without them.

vadz replied:
Sorry, I forgot to answer this one: I thought that perhaps we could avoid `xxxIsSet` by just checking if the colour is valid in `GetXXX()` and returning its default value (used in `ComputeColours()`) if it isn't.

If this doesn't work for some reason I'm missing, it's not a big deal to keep the `xxxIsSet`, of course, but I think they should be renamed to `xxxIsSetByUser` or, maybe, invert its meaning and call it `useDefaultXXX` because currently "is set" is rather confusing: it can be false even if the colour is, actually, set to a valid value.

NewPagodi replied:
I've changed the names to m_useDefaultXXXColour.  That probably is a better name for the purpose those variables serve.

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -1983,433 +1997,743 @@ PRectangle Window::GetMonitorRect(Point pt) {
 
+wxSTCListBox::~wxSTCListBox()
+{
+    m_labels.clear();
```
This is not really needed, why manually do something done automatically anyhow?

NewPagodi replied:
This has been removed.

vadz commented on src/stc/PlatWX.cpp:
```diff
@@ -2418,37 +2742,52 @@ PRectangle ListBoxImpl::GetDesiredRect() {
-    GETLB(wid)->Freeze();
+void wxSTCListBox::SetList(const char* list, char separator, char typesep)
+{
+    Freeze();
```
Using `wxWindowUpdateLocker` is safer.

NewPagodi replied:
This has been changed,

vadz commented on src/stc/PlatWXcocoa.mm:
```diff
@@ -0,0 +1,132 @@
+    wxWindow*        m_wxWin;
+}
+
+- (id)initWithwxWin:(wxWindow*) wxwin;
```
```suggestion
- (id)initWithwxWin:(wxWindow*) wxWin;
```
just for consistency

vadz reviewed (approved):
The promised week turned into a month, but, finally, I did review the code and, as expected, I don't have any real comments about it, just a few nitpicks here and there.

The amount of platform-specific code needed to achieve what you did is truly dispiriting, but I won't even try to understand what could we do to encapsulate the differences in wx because I know I won't have time to do it anyhow (and I suspect neither will you), so we'll just have to live with this.

I can certainly merge it as is, but if you could please address the comments below, it would be great. If you won't be able to do it in the near future, please let me know.

Thanks again for all your work on this!

vadz committed:
Apply suggestions from code review

Co-Authored-By: NewPagodi <NewPagodi@users.noreply.github.com>

Commit:
Apply more suggestions from code review

This commit removes an attempt to initially hide a frame which was
unnecessary since frames are initially hidden by defautlt, removes an
unncecessary destructor that only performed actions that would happen
anyway, and replaces a Freeze/Thaw pair with wxWindowUpdateLocker.

Commit:
Merge branch 'stcpopup' of https://github.com/NewPagodi/wxWidgets into stcpopup

Commit:
Change names of XXXIsSet variables in wxSTCListBoxVisualData

Instead name the variables useDefaultXXXColour since it more accurately
describes what purpose the variables are serving.

vadz commented:
Merged, thanks again for all your work on this!

vadz commented:
@NewPagodi Unfortunately I'm seeing test failures with the tests added here, see e.g. https://travis-ci.org/wxWidgets/wxWidgets/jobs/559896693 but I also can reproduce this locally without Xvfb: both `AutoComp` and `Calltip` tests fail sporadically for me (or, rather, pass sporadically, as they fail more often than not, especially with GTK2, but even with GTK3 I still see the failures often enough too). I've tried debugging this, but without much success...

For now I think I'll just disable these tests under GTK, but it would be nice to fix them, especially if this means that the popup actually doesn't work correctly.

NewPagodi commented:
Well that's frustrating.  GTK+ was the only port where things seemed to work out of the box. I think the location of the mouse clicks must miscalculated and partly random for some reason so that they sometimes landing outside the application window.  I don't think the popup is really taking focus, because if it was, the test would fail 100% of the time.
